libavcodec/hevcdec.c

```diff
@@ -2654,6 +2654,221 @@ static int hls_slice_data_wpp(HEVCContext *s, const H2645NAL *nal)
     return res;
 }
 
+/* Does this CTB abut another tile? */
+static int is_boundary_left_tile(HEVCContext* s, int ctb_addr_ts)
+{ 
+    int ctb_addr_rs = s->ps.pps->ctb_addr_ts_to_rs[ctb_addr_ts];
+
+    return ctb_addr_rs % s->ps.sps->ctb_width > 0 &&
+           s->ps.pps->tile_id[ctb_addr_ts] !=
+           s->ps.pps->tile_id[s->ps.pps->ctb_addr_rs_to_ts[ctb_addr_rs-1]];
+}
+
+static int is_boundary_upper_tile(HEVCContext* s, int ctb_addr_ts)
+{ 
+    int ctb_addr_rs = s->ps.pps->ctb_addr_ts_to_rs[ctb_addr_ts];
+
+    return ctb_addr_rs / s->ps.sps->ctb_width > 0 &&
+           s->ps.pps->tile_id[ctb_addr_ts] !=
+           s->ps.pps->tile_id[s->ps.pps->ctb_addr_rs_to_ts[ctb_addr_rs - s->ps.sps->ctb_width]];
+}
+
+static int hls_decode_entry_tiles(AVCodecContext *avctxt, void *input_tile, int job, int self_id)
+{
+    HEVCContext *s1  = avctxt->priv_data, *s;
+    HEVCLocalContext *lc;
+    int ctb_size    = 1<< s1->ps.sps->log2_ctb_size;
+    int more_data   = 1;
+    int *tile_p    = input_tile;
+    int tile = tile_p[job];
+
+    int slice_start_tile = s1->ps.pps->tile_id[s1->ps.pps->ctb_addr_rs_to_ts[s1->sh.slice_ctb_addr_rs]];
+    int ctb_addr_rs = s1->ps.pps->tile_pos_rs[slice_start_tile + tile];
+    int ctb_addr_ts = s1->ps.pps->ctb_addr_rs_to_ts[ctb_addr_rs];
+    int ret;
+
+    s = s1->sList[self_id];
+    lc = s->HEVClc;
+
+    if (tile) {
+        ret = init_get_bits8(&lc->gb, s->data + s->sh.offset[tile - 1], s->sh.size[tile - 1]);
+        if (ret < 0) {
+            goto error;
+        }
+    }
+
+    while(more_data && ctb_addr_ts < s->ps.sps->ctb_size) {
+        int x_ctb = (ctb_addr_rs % s->ps.sps->ctb_width) << s->ps.sps->log2_ctb_size;
+        int y_ctb = (ctb_addr_rs / s->ps.sps->ctb_width) << s->ps.sps->log2_ctb_size;
+
+        hls_decode_neighbour(s, x_ctb, y_ctb, ctb_addr_ts);
+        ret = ff_hevc_cabac_init(s, ctb_addr_ts);
+        if (ret < 0)
+            goto error;
+
+        hls_sao_param(s, x_ctb >> s->ps.sps->log2_ctb_size, y_ctb >> s->ps.sps->log2_ctb_size);
+        s->deblock[ctb_addr_rs].beta_offset = s->sh.beta_offset;
+        s->deblock[ctb_addr_rs].tc_offset   = s->sh.tc_offset;
+        s->filter_slice_edges[ctb_addr_rs]  = s->sh.slice_loop_filter_across_slices_enabled_flag;
+
+        more_data = hls_coding_quadtree(s, x_ctb, y_ctb, s->ps.sps->log2_ctb_size, 0);
+
+        if (more_data < 0) {
+            ret = more_data;
+            goto error;
+        }
+        ctb_addr_ts++;
+        ff_hevc_save_states(s, ctb_addr_ts);
+
+        if ((x_ctb+ctb_size) >= s->ps.sps->width && (y_ctb+ctb_size) >= s->ps.sps->height) {
+            break;
+        }
+
+        if (s->ps.pps->tile_id[ctb_addr_ts] != s->ps.pps->tile_id[ctb_addr_ts-1]) {
+            break;
+        }
+        
+        ctb_addr_rs = s->ps.pps->ctb_addr_ts_to_rs[ctb_addr_ts];
+    }
+
+    return ctb_addr_ts;
+error:
+    s->tab_slice_address[ctb_addr_rs] = -1;
+    return ret;
+}
+
+static int slicecount = 0;
+
```

Comment:
decode a slice that is split into tiles.  This is similar to hls_slice_data_wpp.